include/swift/Parse/Parser.h

```diff
@@ -411,16 +418,22 @@ class Parser {
     Parser &P;
     ParserPosition PP;
     DiagnosticTransaction DT;
+    llvm::Optional<syntax::SyntaxParsingContext> SynContext;
     bool Backtrack = true;
 
   public:
     BacktrackingScope(Parser &P)
-      : P(P), PP(P.getParserPosition()), DT(P.Diags) {}
+        : P(P), PP(P.getParserPosition()), DT(P.Diags) {
+      SynContext.emplace(P.SyntaxContext);
```

Comment:
We don't need `Optional<>` wrapper here since it's always created. We can use `SynContext(P.SyntaxContext)` in the constructor.

Reply:
`Optional` (and `reset()` in `cancelBacktrack()`) is because we should immediately deconstruct SPC at `cancelBacktrack()`. Consider:
```cpp
ParserResult parseSomething() {
  SourceLoc Start, End;
  Start = consumeToken(tok::start_some);

  BacktrackingScope Backtrack(*this);
  ...
  if (consumeIf(tok::end_some, End)) {
    Backtrack.cancelBacktrack();
    auto Result = new (Context) SomeExpr(Start, End);

    // <HERE>

    return makeParserResult(Result);
  }
  ...
}
```
If we want to insert this at `<HERE>`:
``` swift
    SyntaxContext->createNodeInPlace(SyntaxKind::SomeExpr);
```
It can be problematic without `reset()` in `cancelBacktrack()` because `tok::start_some` token is in the parent context whereas `tok::end_some` is in the SPC in `Backtrack`.

Reply:
I see. This is to say when we cancel backtracking, we should immediately accumulate transparently and switch to the parent context. Could you add some doc-comments there?